README.md
What is this?
=============

Turtle in a Pond is a strategy game. The goal is to surround the turtle before it runs off the screen.

How to use?
===========

Turtle in a Pond is not part of the Sugar desktop, but can be added.  Please refer to;

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
* [How to use Sugar](https://help.sugarlabs.org/),
* [How to use Turtle in a Pond](https://help.sugarlabs.org/turtle_in_a_pond.html)

How to upgrade?
===============

On Sugar desktop systems;
* use [My Settings](https://help.sugarlabs.org/my_settings.html), [Software Update](https://help.sugarlabs.org/my_settings.html#software-update), or;
* use Browse to open [activities.sugarlabs.org](https://activities.sugarlabs.org/), search for `Turtle Pond`, then download.


NEWS
NEWS


10

ENHANCEMENT:

* New translations
* Refactored 'new_dot' function (Sourabha G)
* Renamed variables (Sourabha G)
* disabled new game button when gameover page is shown (Sourabha G)
* Fix source remove warning (Ibiam Chihurumnaya)
* Save and restore 'best-time' from Journal (Sourabha G)
* Feature: gameover page with 'best-time' and 'your-time' (Sourabha G)
* Feature: Stopwatch in each game (Sourabha G)
* Deanglify links to help.sugarlabs.org (James Cameron)
* Port from GObject to Glib (Ayush Nawal)
* Fix README.md for Honey classification (James Cameron)
* Port to Python 3 (Bottersnike)
* fix PyGIWarnings (Ibiam Chihurumnaya)
* Update metadata in activity.info (Vipul Gupta)
* Add README (rdrsadhu)

9

ENHANCEMENT:

* New translations

7

ENHANCEMENT:
* Ignacio Rodríguez port to gtk3

6

ENHANCEMENTS:
* Added play levels (beginner, intermediate, expert, custom)
* Added better end-of-game feedback

5

ENHANCEMENTS:
* Added weighting function to turtle strategy (perhaps it is too difficult
  now?)
* Made dots a bit larger

BUG FIX:
* Added missing space in error messages

4

ENHANCEMENTS:
* More tools for building strategies
* Slightly more challenging default behavior for turtle

BUG FIX:
* Fixed bug with errant button presses

3

ENHANCEMENTS:
* Improved graphics (turtle is now a turtle)
* Error messages from custom strategy appears on toolbar
* Reloading strategy starts new game
* Improved documentation; please see
  http://wiki.sugarlabs.org/go/Activities/Turtle_in_a_Pond

2

BUG FIX:
* Fixed problem launching with old-style toolbars

1

* New project


genxo.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Copyright (c) 2011 Walter Bender

# Port To GTK3:
# Ignacio Rodriguez <ignaciorodriguez@sugarlabs.org>

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA


class SVG:
    ''' SVG generators '''

    def __init__(self):
        self._scale = 1
        self._stroke_width = 1
        self._fill = '#FFFFFF'
        self._stroke = '#000000'

    def _svg_style(self, extras=""):
        return "%s%s%s%s%s%f%s%s%s" % (" style=\"fill:",
                                       self._fill,
                                       ";stroke:",
                                       self._stroke,
                                       ";stroke-width:",
                                       self._stroke_width,
                                       ";",
                                       extras,
                                       "\" />\n")

    def _svg_xo(self):
        self.set_stroke_width(3.5)
        svg_string = "<path d=\"M33.233,35.1l10.102,10.1c0.752,0.75,"
        "1.217,1.783,1.217,2.932   c0,2.287-1.855,4.143-4.146,4.143c-1.145,"
        "0-2.178-0.463-2.932-1.211L27.372,40.961l-10.1,10.1c-0.75,0.75-1.787,"
        "1.211-2.934,1.211   c-2.284,0-4.143-1.854-4.143-4.141c0-1.146,"
        "0.465-2.184,1.212-2.934l10.104-10.102L11.409,24.995   "
        "c-0.747-0.748-1.212-1.785-1.212-2.93c0-2.289,1.854-4.146,"
        "4.146-4.146c1.143,0,2.18,0.465,2.93,1.214l10.099,10.102l10.102-10.103"
        "   c0.754-0.749,1.787-1.214,2.934-1.214c2.289,0,4.146,1.856,"
        "4.146,4.145c0,1.146-0.467,2.18-1.217,2.932L33.233,35.1z\""
        svg_string += self._svg_style()
        svg_string += "\n<circle cx=\"27.371\" cy=\"10.849\" r=\"8.122\""
        svg_string += self._svg_style()
        return svg_string

    def _background(self, scale):
        return self._svg_rect(54.5 * scale, 54.5 * scale, 4, 4, 0.25, 0.25)

    def header(self, scale=1, background=True):
        svg_string = "<?xml version=\"1.0\" encoding=\"UTF-8\""
        svg_string += " standalone=\"no\"?>\n"
        svg_string += "<!-- Created with Emacs -->\n"
        svg_string += "<svg\n"
        svg_string += "   xmlns:svg=\"http://www.w3.org/2000/svg\"\n"
        svg_string += "   xmlns=\"http://www.w3.org/2000/svg\"\n"
        svg_string += "   version=\"1.0\"\n"
        svg_string += "%s%f%s" % ("   width=\"", scale * 55 * self._scale,
                                  "\"\n")
        svg_string += "%s%f%s" % ("   height=\"", scale * 55 * self._scale,
                                  "\">\n")
        svg_string += "%s%f%s%f%s" % ("<g\n       transform=\"matrix(",
                                      self._scale, ",0,0,", self._scale,
                                      ",0,0)\">\n")
        if background:
            svg_string += self._background(scale)
        return svg_string

    def footer(self):
        svg_string = "</g>\n"
        svg_string += "</svg>\n"
        return svg_string

    #
    # Utility functions
    #
    def set_scale(self, scale=1.0):
        self._scale = scale

    def set_colors(self, colors):
        self._stroke = colors[0]
        self._fill = colors[1]

    def set_stroke_width(self, stroke_width=1.0):
        self._stroke_width = stroke_width


def generate_xo(scale=1, colors=["#FFFFFF", "#000000"]):
    svg = SVG()
    svg.set_scale(scale)
    svg.set_colors(colors)
    svg_string = svg.header(background=False)
    svg_string += svg._svg_xo()
    svg_string += svg.footer()
    return svg_string


utils.py
# Copyright (c) 2011 Walter Bender

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA


from gi.repository import GdkPixbuf
from io import StringIO
try:
    OLD_SUGAR_SYSTEM = False
    import json
    json.dumps
    from json import load as jload
    from json import dump as jdump
except (ImportError, AttributeError):
    try:
        import simplejson as json
        from simplejson import load as jload
        from simplejson import dump as jdump
    except BaseException:
        OLD_SUGAR_SYSTEM = True


def json_load(text):
    """ Load JSON data using what ever resources are available. """
    if OLD_SUGAR_SYSTEM is True:
        listdata = json.read(text)
    else:
        # strip out leading and trailing whitespace, nulls, and newlines
        io = StringIO(text)
        try:
            listdata = jload(io)
        except ValueError:
            # assume that text is ascii list
            listdata = text.split()
            for i, value in enumerate(listdata):
                listdata[i] = int(value)
    return listdata


def json_dump(data):
    """ Save data using available JSON tools. """
    if OLD_SUGAR_SYSTEM is True:
        return json.write(data)
    else:
        _io = StringIO()
        jdump(data, _io)
        return _io.getvalue()


def svg_str_to_pixbuf(svg_string):
    ''' Load pixbuf from SVG string '''
    pl = GdkPixbuf.PixbufLoader.new_with_type('svg')
    pl.write(svg_string)
    pl.close()
    pixbuf = pl.get_pixbuf()
    return pixbuf


sprites.py
# -*- coding: utf-8 -*-

# Copyright (c) 2007-8, Playful Invention Company.
# Copyright (c) 2008-11 Walter Bender

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

'''

sprites.py is a simple sprites library for managing graphics objects,
'sprites', on a Gtk.DrawingArea. It manages multiple sprites with
methods such as move, hide, set_layer, etc.

There are two classes:

class Sprites maintains a collection of sprites
class Sprite manages individual sprites within the collection.

Example usage:
        # Import the classes into your program.
        from sprites import Sprites Sprite

        # Create a new sprite collection associated with your widget
        self.sprite_list = Sprites(widget)

        # Create a "pixbuf" (in this example, from SVG).
        my_pixbuf = svg_str_to_pixbuf("<svg>...some svg code...</svg>")

        # Create a sprite at position x1, y1.
        my_sprite = sprites.Sprite(self.sprite_list, x1, y1, my_pixbuf)

        # Move the sprite to a new position.
        my_sprite.move((x1+dx, y1+dy))

        # Create another "pixbuf".
        your_pixbuf = svg_str_to_pixbuf("<svg>...some svg code...</svg>")

        # Create a sprite at position x2, y2.
        your_sprite = sprites.Sprite(self.sprite_list, x2, y2, my_pixbuf)

        # Assign the sprites to layers.
        # In this example, your_sprite will be on top of my_sprite.
        my_sprite.set_layer(100)
        your_sprite.set_layer(200)

        # Now put my_sprite on top of your_sprite.
        my_sprite.set_layer(300)

        cr = self.window.cairo_create()
        # In your activity's do_expose_event, put in a call to redraw_sprites
        self.sprites.redraw_sprites(event.area, cairo_context)

# method for converting SVG to a gtk pixbuf
def svg_str_to_pixbuf(svg_string):
    pl = GdkPixbuf.PixbufLoader('svg')
    pl.write(svg_string)
    pl.close()
    pixbuf = pl.get_pixbuf()
    return pixbuf

'''

import gi
gi.require_version('PangoCairo', '1.0')

from gi.repository import GdkPixbuf, Gdk
from gi.repository import Pango, PangoCairo
import cairo


class Sprites:
    ''' A class for the list of sprites and everything they share in common '''

    def __init__(self, widget):
        ''' Initialize an empty array of sprites '''
        self.cr = None
        self.widget = widget
        self.list = []

    def set_cairo_context(self, cr):
        ''' Cairo context may be set or reset after __init__ '''
        self.cr = cr

    def get_sprite(self, i):
        ''' Return a sprint from the array '''
        if i < 0 or i > len(self.list) - 1:
            return(None)
        else:
            return(self.list[i])

    def length_of_list(self):
        ''' How many sprites are there? '''
        return(len(self.list))

    def append_to_list(self, spr):
        ''' Append a new sprite to the end of the list. '''
        self.list.append(spr)

    def insert_in_list(self, spr, i):
        ''' Insert a sprite at position i. '''
        if i < 0:
            self.list.insert(0, spr)
        elif i > len(self.list) - 1:
            self.list.append(spr)
        else:
            self.list.insert(i, spr)

    def remove_from_list(self, spr):
        ''' Remove a sprite from the list. '''
        if spr in self.list:
            self.list.remove(spr)

    def find_sprite(self, pos, inverse=False):
        ''' Search based on (x, y) position. Return the 'top/first' one. '''
        list = self.list[:]
        if not inverse:
            list.reverse()
        for spr in list:
            if spr.hit(pos):
                return spr
        return None

    def redraw_sprites(self, area=None, cr=None):
        ''' Redraw the sprites that intersect area. '''
        # I think I need to do this to save Cairo some work
        if cr is None:
            cr = self.cr
        else:
            self.cr = cr
        if cr is None:
            print('sprites.redraw_sprites: no Cairo context')
            return
        for spr in self.list:
            if area is None:
                spr.draw(cr=cr)
            else:
                intersection = spr.rect.intersect(area)
                if intersection.width > 0 or intersection.height > 0:
                    spr.draw(cr=cr)


class Sprite:
    ''' A class for the individual sprites '''

    def __init__(self, sprites, x, y, image):
        ''' Initialize an individual sprite '''
        self._sprites = sprites
        self.save_xy = (x, y)  # remember initial (x, y) position
        self.rect = [int(x), int(y), 0, 0]
        self._scale = [12]
        self._rescale = [True]
        self._horiz_align = ["center"]
        self._vert_align = ["middle"]
        self._x_pos = [None]
        self._y_pos = [None]
        self._fd = None
        self._bold = False
        self._italic = False
        self._color = None
        self._margins = [0, 0, 0, 0]
        self.layer = 100
        self.labels = []
        self.images = []
        self._dx = []  # image offsets
        self._dy = []
        self.type = None
        self.set_image(image)
        self._sprites.append_to_list(self)

    def set_image(self, image, i=0, dx=0, dy=0):
        ''' Add an image to the sprite. '''
        while len(self.images) < i + 1:
            self.images.append(None)
            self._dx.append(0)
            self._dy.append(0)
        self.images[i] = image
        self._dx[i] = dx
        self._dy[i] = dy
        if hasattr(self.images[i], 'get_width'):
            w = self.images[i].get_width()
            h = self.images[i].get_height()
        else:
            w, h = self.images[i].get_size()
        if i == 0:  # Always reset width and height when base image changes.
            self.rect[2] = w + dx
            self.rect[3] = h + dy
        else:
            if w + dx > self.rect[2]:
                self.rect[2] = w + dx
            if h + dy > self.rect[3]:
                self.rect[3] = h + dy

    def move(self, pos):
        ''' Move to new (x, y) position '''
        self.inval()
        self.rect[0], self.rect[1] = int(pos[0]), int(pos[1])
        self.inval()

    def move_relative(self, pos):
        ''' Move to new (x+dx, y+dy) position '''
        self.inval()
        self.rect[0] += int(pos[0])
        self.rect[1] += int(pos[1])
        self.inval()

    def get_xy(self):
        ''' Return current (x, y) position '''
        return (self.rect[0], self.rect[1])

    def get_dimensions(self):
        ''' Return current size '''
        return (self.rect[2], self.rect[3])

    def get_layer(self):
        ''' Return current layer '''
        return self.layer

    def set_shape(self, image, i=0):
        ''' Set the current image associated with the sprite '''
        self.inval()
        self.set_image(image, i)
        self.inval()

    def set_layer(self, layer=None):
        ''' Set the layer for a sprite '''
        self._sprites.remove_from_list(self)
        if layer is not None:
            self.layer = layer
        for i in range(self._sprites.length_of_list()):
            if self.layer < self._sprites.get_sprite(i).layer:
                self._sprites.insert_in_list(self, i)
                self.inval()
                return
        self._sprites.append_to_list(self)
        self.inval()

    def set_label(self, new_label, i=0):
        ''' Set the label drawn on the sprite '''
        self._extend_labels_array(i)
        if isinstance(new_label, str):
            # pango doesn't like nulls
            self.labels[i] = new_label.replace("\0", " ")
        else:
            self.labels[i] = str(new_label)
        self.inval()

    def set_margins(self, w=0, t=0, r=0, b=0):
        ''' Set the margins for drawing the label '''
        self._margins = [w, t, r, b]

    def _extend_labels_array(self, i):
        ''' Append to the labels attribute list '''
        if self._fd is None:
            self.set_font('Sans')
        if self._color is None:
            self._color = (0., 0., 0.)
        while len(self.labels) < i + 1:
            self.labels.append(" ")
            self._scale.append(self._scale[0])
            self._rescale.append(self._rescale[0])
            self._horiz_align.append(self._horiz_align[0])
            self._vert_align.append(self._vert_align[0])
            self._x_pos.append(self._x_pos[0])
            self._y_pos.append(self._y_pos[0])

    def set_font(self, font):
        ''' Set the font for a label '''
        self._fd = Pango.FontDescription(font)

    def set_label_color(self, rgb):
        ''' Set the font color for a label '''
        COLORTABLE = {'black': '#000000', 'white': '#FFFFFF',
                      'red': '#FF0000', 'yellow': '#FFFF00',
                      'green': '#00FF00', 'cyan': '#00FFFF',
                      'blue': '#0000FF', 'purple': '#FF00FF',
                      'gray': '#808080'}
        if rgb.lower() in COLORTABLE:
            rgb = COLORTABLE[rgb.lower()]
        # Convert from '#RRGGBB' to floats
        self._color = (int('0x' + rgb[1:3], 16) / 256.,
                       int('0x' + rgb[3:5], 16) / 256.,
                       int('0x' + rgb[5:7], 16) / 256.)
        return

    def set_label_attributes(self, scale, rescale=True, horiz_align="center",
                             vert_align="middle", x_pos=None, y_pos=None, i=0):
        ''' Set the various label attributes '''
        self._extend_labels_array(i)
        self._scale[i] = scale
        self._rescale[i] = rescale
        self._horiz_align[i] = horiz_align
        self._vert_align[i] = vert_align
        self._x_pos[i] = x_pos
        self._y_pos[i] = y_pos

    def hide(self):
        ''' Hide a sprite '''
        self.inval()
        self._sprites.remove_from_list(self)

    def restore(self):
        ''' Restore a hidden sprite '''
        self.set_layer()

    def inval(self):
        ''' Invalidate a region for gtk '''
        # self._sprites.window.invalidate_rect(self.rect, False)
        self._sprites.widget.queue_draw_area(self.rect[0],
                                             self.rect[1],
                                             self.rect[2],
                                             self.rect[3])

    def draw(self, cr=None):
        ''' Draw the sprite (and label) '''
        if cr is None:
            cr = self._sprites.cr
        if cr is None:
            print('sprite.draw: no Cairo context.')
            return
        for i, img in enumerate(self.images):
            if isinstance(img, GdkPixbuf.Pixbuf):
                Gdk.cairo_set_source_pixbuf(cr, img,
                                            self.rect[0] + self._dx[i],
                                            self.rect[1] + self._dy[i])
                cr.rectangle(self.rect[0] + self._dx[i],
                             self.rect[1] + self._dy[i],
                             self.rect[2],
                             self.rect[3])
                cr.fill()
            elif isinstance(img, cairo.ImageSurface):
                cr.set_source_surface(img, self.rect[0] + self._dx[i],
                                      self.rect[1] + self._dy[i])
                cr.rectangle(self.rect[0] + self._dx[i],
                             self.rect[1] + self._dy[i],
                             self.rect[2],
                             self.rect[3])
                cr.fill()
            else:
                print('sprite.draw: source not pixbuf ({})'.format(type(img)))
        if len(self.labels) > 0:
            self.draw_label(cr)

    def hit(self, pos):
        ''' Is (x, y) on top of the sprite? '''
        x, y = pos
        if x < self.rect[0]:
            return False
        if x > self.rect[0] + self.rect[2]:
            return False
        if y < self.rect[1]:
            return False
        if y > self.rect[1] + self.rect[3]:
            return False
        return True

    def draw_label(self, cr):
        ''' Draw the label based on its attributes '''
        my_width = self.rect[2] - self._margins[0] - self._margins[2]
        if my_width < 0:
            my_width = 0
        my_height = self.rect[3] - self._margins[1] - self._margins[3]
        for i in range(len(self.labels)):
            pl = PangoCairo.create_layout(cr)
            pl.set_text(str(self.labels[i]), -1)
            self._fd.set_size(int(self._scale[i] * Pango.SCALE))
            pl.set_font_description(self._fd)
            w = pl.get_size()[0] / Pango.SCALE
            if w > my_width:
                if self._rescale[i]:
                    self._fd.set_size(
                        int(self._scale[i] * Pango.SCALE * my_width / w))
                    pl.set_font_description(self._fd)
                    w = pl.get_size()[0] / Pango.SCALE
                else:
                    j = len(self.labels[i]) - 1
                    while(w > my_width and j > 0):
                        pl.set_text(
                            "…" + self.labels[i][len(self.labels[i]) - j:], -1)
                        self._fd.set_size(int(self._scale[i] * Pango.SCALE))
                        pl.set_font_description(self._fd)
                        w = pl.get_size()[0] / Pango.SCALE
                        j -= 1
            if self._x_pos[i] is not None:
                x = int(self.rect[0] + self._x_pos[i])
            elif self._horiz_align[i] == "center":
                x = int(self.rect[0] + self._margins[0] + (my_width - w) / 2)
            elif self._horiz_align[i] == 'left':
                x = int(self.rect[0] + self._margins[0])
            else:  # right
                x = int(self.rect[0] + self.rect[2] - w - self._margins[2])
            h = pl.get_size()[1] / Pango.SCALE
            if self._y_pos[i] is not None:
                y = int(self.rect[1] + self._y_pos[i])
            elif self._vert_align[i] == "middle":
                y = int(self.rect[1] + self._margins[1] + (my_height - h) / 2)
            elif self._vert_align[i] == "top":
                y = int(self.rect[1] + self._margins[1])
            else:  # bottom
                y = int(self.rect[1] + self.rect[3] - h - self._margins[3])
            cr.save()
            cr.translate(x, y)
            cr.set_source_rgb(self._color[0], self._color[1], self._color[2])
            PangoCairo.update_layout(cr, pl)
            PangoCairo.show_layout(cr, pl)
            cr.restore()

    def label_width(self, cr=None):
        ''' Calculate the width of a label '''
        if cr is None:
            cr = self._sprites.cr
        max = 0
        for i in range(len(self.labels)):
            pl = PangoCairo.create_layout(cr)
            pl.set_text(str(self.labels[i]), -1)
            self._fd.set_size(int(self._scale[i] * Pango.SCALE))
            pl.set_font_description(self._fd)
            w = pl.get_size()[0] / Pango.SCALE
            if w > max:
                max = w
        return max

    def label_safe_width(self):
        ''' Return maximum width for a label '''
        return self.rect[2] - self._margins[0] - self._margins[2]

    def label_safe_height(self):
        ''' Return maximum height for a label '''
        return self.rect[3] - self._margins[1] - self._margins[3]

    def label_left_top(self):
        ''' Return the upper-left corner of the label safe zone '''
        return(self._margins[0], self._margins[1])

    def get_pixel(self, pos, i=0):
        ''' Return the pixl at (x, y) '''
        x, y = pos
        x = x - self.rect[0]
        y = y - self.rect[1]
        if y > self.images[i].get_height() - 1:
            return(-1, -1, -1, -1)
        try:
            array = self.images[i].get_pixels()
            if array is not None:
                offset = (y * self.images[i].get_width() + x) * 4
                r, g, b, a = ord(array[offset]), ord(array[offset + 1]),\
                    ord(array[offset + 2]), ord(array[offset + 3])
                return(r, g, b, a)
            else:
                return(-1, -1, -1, -1)
        except IndexError:
            print("Index Error: {} {}".format(len(array), offset))
            return(-1, -1, -1, -1)


TurtlePondActivity.py
#Copyright (c) 2011 Walter Bender
# Port To GTK3:
# Ignacio Rodriguez <ignaciorodriguez@sugarlabs.org>
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA

import gi
gi.require_version('Gtk', '3.0')

from gi.repository import Gtk, Gdk

import sugar3
from sugar3.activity import activity
from sugar3 import profile
from sugar3.graphics.toolbarbox import ToolbarBox
from sugar3.bundle.activitybundle import ActivityBundle
from sugar3.activity.widgets import ActivityToolbarButton
from sugar3.activity.widgets import StopButton
from sugar3.graphics.toolbarbox import ToolbarButton

from sugar3.graphics.toolbutton import ToolButton
from sugar3.graphics.menuitem import MenuItem
from sugar3.graphics.icon import Icon
from sugar3.datastore import datastore
from sugar3.graphics.objectchooser import ObjectChooser

from toolbar_utils import button_factory, image_factory, label_factory, \
    separator_factory, radio_factory

from gettext import gettext as _
import locale
import os.path

from game import Game
from genxo import generate_xo
from utils import svg_str_to_pixbuf

import logging
_logger = logging.getLogger('turtle-in-a-pond-activity')

BEGINNER = 0
INTERMEDIATE = 1
EXPERT = 2
CUSTOM = 3
LEVEL_LABELS = [_('Beginner'), _('Intermediate'), _('Expert'),
                _('My strategy')]


class TurtlePondActivity(activity.Activity):
    """ Turtle in a Pond puzzle game """

    def __init__(self, handle):
        """ Initialize the toolbars and the game board """
        super(TurtlePondActivity, self).__init__(handle)
        self.nick = profile.get_nick_name()
        if profile.get_color() is not None:
            self.colors = profile.get_color().to_string().split(',')
        else:
            self.colors = ['#A0FFA0', '#FF8080']

        self._setup_toolbars()

        # Create a canvas
        canvas = Gtk.DrawingArea()
        canvas.set_size_request(Gdk.Screen.width(), \
                                Gdk.Screen.height())
        self.set_canvas(canvas)
        canvas.show()
        self.show_all()

        self._game = Game(canvas, parent=self, colors=self.colors)

        # TODO: Restore game state from Journal or start new game
        self._game.new_game()

    def _setup_toolbars(self):
        """ Setup the toolbars. """

        self.max_participants = 1

        toolbox = ToolbarBox()

        # Activity toolbar
        activity_button = ActivityToolbarButton(self)

        toolbox.toolbar.insert(activity_button, 0)
        activity_button.show()

        self.set_toolbar_box(toolbox)
        toolbox.show()
        self.toolbar = toolbox.toolbar

        self._new_game_button = button_factory(
            'new-game', self.toolbar, self._new_game_cb,
            tooltip=_('Start a new game.'))

        separator_factory(self.toolbar, False, True)

        self.beginner_button = radio_factory(
            'beginner',
            self.toolbar,
            self._level_cb,
            cb_arg=BEGINNER,
            tooltip=LEVEL_LABELS[BEGINNER],
            group=None)
        self.intermediate_button = radio_factory(
            'intermediate',
            self.toolbar,
            self._level_cb,
            cb_arg=INTERMEDIATE,
            tooltip=LEVEL_LABELS[INTERMEDIATE],
            group=self.beginner_button)
        self.expert_button = radio_factory(
            'expert',
            self.toolbar,
            self._level_cb,
            cb_arg=EXPERT,
            tooltip=LEVEL_LABELS[EXPERT],
            group=self.beginner_button)
        self.custom_button = radio_factory(
            'view-source',
            self.toolbar,
            self._level_cb,
            cb_arg=CUSTOM,
            tooltip=LEVEL_LABELS[CUSTOM],
            group=self.beginner_button)

        self.status = label_factory(self.toolbar, '')

        separator_factory(toolbox.toolbar, True, False)

        self.load_python = button_factory(
            'pippy-openoff', self.toolbar,
            self._do_load_python_cb,
            tooltip=_('Load strategy from Journal'))

        stop_button = StopButton(self)
        stop_button.props.accelerator = '<Ctrl>q'
        toolbox.toolbar.insert(stop_button, -1)
        stop_button.show()

    def _level_cb(self, button, level):
        if level == CUSTOM and self._game.strategies[CUSTOM] is None:
            level = EXPERT
            self.expert_button.set_active(True)
        self._game.level = level
        self._game.new_game()

    def _new_game_cb(self, button=None):
        ''' Start a new game. '''
        if not self._game.gameover_flag:
            self._game.new_game()
        else: 
            self._new_game_button.props.sensitive = False
        self._new_game_button.props.sensitive = True

    def write_file(self, file_path):
        """ Write the grid status to the Journal """
        return

    def _restore(self):
        """ Restore the game state from metadata """
        return

    def _do_load_python_cb(self, button):
        ''' Load Python code from the Journal. '''
        self._chooser('org.laptop.Pippy',
                self._load_python_code_from_journal)
        if self._game.strategies[CUSTOM] is not None:
            self.custom_button.set_active(True)
        self._game.level = CUSTOM
        self._game.new_game()

    def _load_python_code_from_journal(self, dsobject):
        """ Read the Python code from the Journal object """
        python_code = None
        try:
            _logger.debug("opening %s " % dsobject.file_path)
            file_handle = open(dsobject.file_path, "r")
            python_code = file_handle.read()
            file_handle.close()
        except IOError:
            _logger.debug("couldn't open %s" % dsobject.file_path)
        self._game.strategies[CUSTOM] = python_code

    def _chooser(self, filter, action):
        ''' Choose an object from the datastore and take some action '''
        chooser = None
        try:
            chooser = ObjectChooser(parent=self, what_filter=filter)
        except TypeError:
            chooser = ObjectChooser(
                None, self,
                Gtk.DialogFlags.MODAL | Gtk.DialogFlags.DESTROY_WITH_PARENT)
        if chooser is not None:
            try:
                result = chooser.run()
                if result == Gtk.ResponseType.ACCEPT:
                    dsobject = chooser.get_selected_object()
                    action(dsobject)
                    dsobject.destroy()
            finally:
                chooser.destroy()
                del chooser
